pkg/autoscaler/autoscaler/util.go

```diff
@@ -126,13 +136,151 @@ func defaultTAC(tac *v1alpha1.TidbClusterAutoScaler) {
 		if len(tac.Spec.TiKV.Metrics) == 0 {
 			tac.Spec.TiKV.Metrics = append(tac.Spec.TiKV.Metrics, defaultMetricSpec)
 		}
+		if tac.Spec.TiKV.ScaleInThreshold == nil {
+			tac.Spec.TiKV.ScaleInThreshold = pointer.Int32Ptr(3)
+		}
+		if tac.Spec.TiKV.ScaleOutThreshold == nil {
+			tac.Spec.TiKV.ScaleOutThreshold = pointer.Int32Ptr(3)
+		}
+		if tac.Spec.TiKV.MetricsTimeDuration == nil {
+			tac.Spec.TiKV.MetricsTimeDuration = pointer.StringPtr("3m")
+		}
+		if tac.Spec.TiKV.ScaleOutIntervalSeconds == nil {
+			tac.Spec.TiKV.ScaleOutIntervalSeconds = pointer.Int32Ptr(300)
+		}
+		if tac.Spec.TiKV.ScaleInIntervalSeconds == nil {
+			tac.Spec.TiKV.ScaleInIntervalSeconds = pointer.Int32Ptr(300)
+		}
 	}
+
 	if tac.Spec.TiDB != nil {
 		if tac.Spec.TiDB.MinReplicas == nil {
 			tac.Spec.TiDB.MinReplicas = pointer.Int32Ptr(1)
 		}
 		if len(tac.Spec.TiDB.Metrics) == 0 {
 			tac.Spec.TiDB.Metrics = append(tac.Spec.TiDB.Metrics, defaultMetricSpec)
 		}
+		if tac.Spec.TiDB.ScaleInThreshold == nil {
+			tac.Spec.TiDB.ScaleInThreshold = pointer.Int32Ptr(3)
+		}
+		if tac.Spec.TiDB.ScaleOutThreshold == nil {
+			tac.Spec.TiDB.ScaleOutThreshold = pointer.Int32Ptr(3)
+		}
+		if tac.Spec.TiDB.MetricsTimeDuration == nil {
+			tac.Spec.TiDB.MetricsTimeDuration = pointer.StringPtr("3m")
+		}
+		if tac.Spec.TiDB.ScaleOutIntervalSeconds == nil {
+			tac.Spec.TiDB.ScaleOutIntervalSeconds = pointer.Int32Ptr(300)
+		}
+		if tac.Spec.TiDB.ScaleInIntervalSeconds == nil {
+			tac.Spec.TiDB.ScaleInIntervalSeconds = pointer.Int32Ptr(300)
+		}
+	}
+}
+
+// updateConsecutiveCount would update the tc annotation depended by the given replicas in each reconciling
+func updateConsecutiveCount(tc *v1alpha1.TidbCluster, tac *v1alpha1.TidbClusterAutoScaler,
+	memberType v1alpha1.MemberType, currentReplicas int32, recommendedReplicas int32) error {
+	if tc.Annotations == nil {
+		tc.Annotations = map[string]string{}
 	}
+
+	targetScaleOutAnn := fmt.Sprintf("%s.%s", memberType.String(), annScaleOutSuffix)
+	targetScaleInAnn := fmt.Sprintf("%s.%s", memberType.String(), annScaleInSuffix)
+
+	var scaleOutCount int
+	var scaleInCount int
+	scaleOutCount, scaleInCount = 0, 0
+	var err error
+
+	if v, existed := tc.Annotations[targetScaleOutAnn]; existed {
```

Comment:
If store the count in annotation, e.g. the threshold is 3, after it increases to 2, operator is down, then after 10 minutes, operator is up and then the result is scaling (even if the resource utilization is below the resource usage threshold during the 10 minutes that the operaotor is down), then the scaling will be done.

Reply:
This is a good catch, I suggest to record is as an issue in order to guarantee the production-ready stability.  For an alpha feature, the first thing we need to ensure is the available for auto-scaler.

Reply:
issued in https://github.com/pingcap/tidb-operator/issues/1651